In [1]:
# Required Python Machine learning Packages
import pandas as pd
import numpy as np

In [2]:
#import data file
iris = pd.read_csv("/content/iris.csv")

In [3]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [4]:
print(iris.dtypes)
print(iris.shape)
print(iris.describe())

sepal_length    float64
sepal_width     float64
petal_length    float64
petal_width     float64
species          object
dtype: object
(150, 5)
       sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000


In [5]:
# for see the null value
iris.isnull().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

In [6]:
#mapping
iris['species']=iris['species'].map({'Iris-setosa':0,'Iris-versicolor':1,'Iris-virginica':2})

In [7]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width  species
0           5.1          3.5           1.4          0.2        0
1           4.9          3.0           1.4          0.2        0
2           4.7          3.2           1.3          0.2        0
3           4.6          3.1           1.5          0.2        0
4           5.0          3.6           1.4          0.2        0

In [8]:
#split dependent and independent variable
X = iris.values[:,0:-1] #select all columns only last column is not consider,
Y = iris.values[:,-1] # select only last column,

In [9]:
#feature scaling techniques
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X)
X = scaler.transform(X)
print(X) 

[[-9.00681170e-01  1.03205722e+00 -1.34127240e+00 -1.31297673e+00]
 [-1.14301691e+00 -1.24957601e-01 -1.34127240e+00 -1.31297673e+00]
 [-1.38535265e+00  3.37848329e-01 -1.39813811e+00 -1.31297673e+00]
 [-1.50652052e+00  1.06445364e-01 -1.28440670e+00 -1.31297673e+00]
 [-1.02184904e+00  1.26346019e+00 -1.34127240e+00 -1.31297673e+00]
 [-5.37177559e-01  1.95766909e+00 -1.17067529e+00 -1.05003079e+00]
 [-1.50652052e+00  8.00654259e-01 -1.34127240e+00 -1.18150376e+00]
 [-1.02184904e+00  8.00654259e-01 -1.28440670e+00 -1.31297673e+00]
 [-1.74885626e+00 -3.56360566e-01 -1.34127240e+00 -1.31297673e+00]
 [-1.14301691e+00  1.06445364e-01 -1.28440670e+00 -1.44444970e+00]
 [-5.37177559e-01  1.49486315e+00 -1.28440670e+00 -1.31297673e+00]
 [-1.26418478e+00  8.00654259e-01 -1.22754100e+00 -1.31297673e+00]
 [-1.26418478e+00 -1.24957601e-01 -1.34127240e+00 -1.44444970e+00]
 [-1.87002413e+00 -1.24957601e-01 -1.51186952e+00 -1.44444970e+00]
 [-5.25060772e-02  2.18907205e+00 -1.45500381e+00 -1.31297673e

In [10]:
#split the data in train and test by 80% and 20%

from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2,
random_state=1)

In [96]:
#model building for logistic regression.
from sklearn.linear_model import LogisticRegression
LogisticRegression = LogisticRegression(penalty='l2', max_iter=50 ) #for penality in that case only support 'l2'
LogisticRegression.fit(X_train, Y_train)
Y_pred=LogisticRegression.predict(X_test)
print(Y_pred)

[0. 1. 1. 0. 2. 1. 2. 0. 0. 2. 1. 0. 2. 1. 1. 0. 1. 1. 0. 0. 1. 1. 2. 0.
 2. 1. 0. 0. 1. 2.]


In [94]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[11  0  0]
 [ 0 12  1]
 [ 0  0  6]]
Classification report: 
              precision    recall  f1-score   support

         0.0       1.00      1.00      1.00        11
         1.0       1.00      0.92      0.96        13
         2.0       0.86      1.00      0.92         6

    accuracy                           0.97        30
   macro avg       0.95      0.97      0.96        30
weighted avg       0.97      0.97      0.97        30

Accuracy of the model:  0.9666666666666667


for random state = 45, Accuracy of the model:  0.9666666666666667
        
for random state = 10, Accuracy of the model:  1.0

for random state = 1, Accuracy of the model:  0.9666666666666667

In [99]:
# use cross validation


from sklearn.model_selection import KFold
kfold_cv = KFold(n_splits = 6, random_state = 1,shuffle=True)

from sklearn.model_selection import cross_val_score

kfold_cv_result = cross_val_score(estimator = LogisticRegression, X=X_train, y = Y_train, cv = kfold_cv)

print(kfold_cv_result)
print(kfold_cv_result.mean())

[1.   1.   0.95 1.   0.95 0.9 ]
0.9666666666666668


when we use cross validation for n_splits = 6 then gives same accuracy score that means our model is not overfited

In [14]:
#this code use when we get difference in above cross validation result and original algorithm 

for train_value, test_value in kfold_cv.split(X_train):
  classifier.fit(X_train[train_value],Y_train[train_value]).predict(X_train[test_value])
Y_pred=classifier.predict(X_test)

In [15]:
acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

Accuracy of the model:  0.9333333333333333


=====Random forest 1=====

In [16]:
from sklearn.ensemble import RandomForestClassifier


Random_Forest_Classifier=RandomForestClassifier(n_estimators=30, criterion='gini', max_depth=30, min_samples_leaf=5 )

#fit the model on the data and predict the values
Random_Forest_Classifier.fit(X_train,Y_train)

Y_pred=Random_Forest_Classifier.predict(X_test)


from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
#confusion matrix
#print(confusion_matrix(Y_test,Y_pred))
print(accuracy_score(Y_test,Y_pred))
#print(classification_report(Y_test,Y_pred))

0.9666666666666667


=====Random forest 2=====

In [17]:

from sklearn.ensemble import RandomForestRegressor

Random_Forest_Regressor=RandomForestRegressor(n_estimators=20, criterion='squared_error', max_depth=25, min_samples_leaf=5 )

Random_Forest_Regressor.fit(X_train,Y_train)

Y_pred=Random_Forest_Regressor.predict(X_test)


from sklearn.metrics import accuracy_score

from sklearn import metrics

print('Mean Absolute Error:', metrics.mean_absolute_error(Y_test, Y_pred))
print('Mean Squared Error:', metrics.mean_squared_error(Y_test, Y_pred))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(Y_test, Y_pred)))

Mean Absolute Error: 0.049485449735449744
Mean Squared Error: 0.029441897199950332
Root Mean Squared Error: 0.17158641321488813


THIS PROBLEM IS CLASSIFICATION PROBLEM THEREFOR REGRESSOR GIVES BAD ACCURACY COMPARE TO CLASSIFIER

====== Gradient Boosted Trees=======

In [18]:
from sklearn.ensemble import GradientBoostingClassifier

# loss='deviance', learning_rate=0.1, n_estimators=(67,89), subsample=(1,2),max_depth=(5,7), n_iter_no_change=(20,40)
#but subsample must be in (0,1]
GradientBoostedTrees = GradientBoostingClassifier( loss='deviance', learning_rate=0.1, n_estimators=89, subsample=1, max_depth=7, n_iter_no_change=40 )

GradientBoostedTrees.fit(X_train, Y_train)
Y_pred=GradientBoostedTrees.predict(X_test)


from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
#confusion matrix
#print(confusion_matrix(Y_test,Y_pred))
print(accuracy_score(Y_test,Y_pred))
#print(classification_report(Y_test,Y_pred))

0.9666666666666667


======Decsion Tree=======

In [19]:
from sklearn.tree import DecisionTreeClassifier

#Decision_Tree = DecisionTreeClassifier(  criterion='entropy', max_depth=(4,7),splitter='best', min_samples_leaf=(6,12))

Decision_Tree = DecisionTreeClassifier(  criterion='entropy', max_depth=7,splitter='best', min_samples_leaf=12)

Decision_Tree.fit(X_train, Y_train)
Y_pred=Decision_Tree.predict(X_test)


from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
#confusion matrix
#print(confusion_matrix(Y_test,Y_pred))
print(accuracy_score(Y_test,Y_pred))
#print(classification_report(Y_test,Y_pred))

0.9666666666666667


=======SVM=======

In [27]:
from sklearn import svm

Support_Vector_Machine = svm.SVC( C=500, kernel='sigmoid', degree=3,gamma=0.1, coef0=0.0, shrinking=True, probability=False, tol=7,  max_iter=7) 
# gamma=0.01,

Support_Vector_Machine.fit(X_train, Y_train)
Y_pred=Support_Vector_Machine.predict(X_test)


from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
#confusion matrix
#print(confusion_matrix(Y_test,Y_pred))
print(accuracy_score(Y_test,Y_pred))
#print(classification_report(Y_test,Y_pred))

0.2


=============KNN==============

In [33]:
from sklearn import neighbors

KNN=neighbors.KNeighborsClassifier(n_neighbors=20, algorithm='auto', p=1)
KNN.fit(X_train, Y_train)
Y_pred=KNN.predict(X_test)


from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
#confusion matrix
#print(confusion_matrix(Y_test,Y_pred))
print(accuracy_score(Y_test,Y_pred))
#print(classification_report(Y_test,Y_pred))

0.9666666666666667


=============ANN==========

In [ ]:
!pip install tensorflow-gpu

In [35]:
import tensorflow as tf

In [36]:
## lets create the ANN
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import LeakyReLU,PReLU,ELU,ReLU
from tensorflow.keras.layers import Dropout

In [51]:
Neural_network = Sequential()
Neural_network.add(Dense(units=67,activation = 'ReLU'))
Neural_network.add(Dense(units=89, activation = 'ReLU'))
#output layer
Neural_network.add(Dense(units=1, activation = 'sigmoid'))
Neural_network.compile(optimizer = 'adam', loss='binary_crossentropy', metrics=['accuracy'])

In [65]:
import tensorflow
opt=tensorflow.keras.optimizers.Adam(learning_rate=0.001)

early_stopping = tf.keras.callbacks.EarlyStopping(
    monitor="val_loss",
    min_delta=0.000,
    patience=20,
    verbose=1,
    mode="auto",
    baseline=None,
    restore_best_weights=False,
)

In [68]:
model_history=Neural_network.fit(X_train,Y_train,validation_split=0.4,batch_size=20,epochs=50, callbacks=early_stopping)

Epoch 1/50
4/4 [==============================] - 0s 25ms/step - loss: -1046999.7500 - accuracy: 0.5694 - val_loss: -538087.6250 - val_accuracy: 0.7292
Epoch 2/50
4/4 [==============================] - 0s 17ms/step - loss: -1049521.8750 - accuracy: 0.5694 - val_loss: -539388.2500 - val_accuracy: 0.7292
Epoch 3/50
4/4 [==============================] - 0s 17ms/step - loss: -1052106.6250 - accuracy: 0.5694 - val_loss: -540699.0000 - val_accuracy: 0.7292
Epoch 4/50
4/4 [==============================] - 0s 11ms/step - loss: -1054498.3750 - accuracy: 0.5694 - val_loss: -542035.1875 - val_accuracy: 0.7292
Epoch 5/50
4/4 [==============================] - 0s 17ms/step - loss: -1057130.0000 - accuracy: 0.5694 - val_loss: -543348.3125 - val_accuracy: 0.7292
Epoch 6/50
4/4 [==============================] - 0s 12ms/step - loss: -1059671.5000 - accuracy: 0.5694 - val_loss: -544665.0000 - val_accuracy: 0.7292
Epoch 7/50
4/4 [==============================] - 0s 17ms/step - loss: -1062414.3750 - a

In [71]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
# Predicting the Test set results
y_pred = Neural_network.predict(X_test)

1/1 [==============================] - 0s 94ms/step


In [85]:
# Calculate the Accuracy
y_pred = (y_pred > 0.5)
from sklearn.metrics import accuracy_score
score=accuracy_score(y_pred,Y_test)
score

0.8

In [86]:
print("Accuracy score for Neural network is  ",score)

Accuracy score for Neural network is   0.8
